Create Task-bot

MerlinLim opened the pull request:
Basic functions of a task bot.
1. Use todo, deadline and event commands to create a task
2. Use List to view current tasks
3. Use done mark tasks as done.
4. Use bye to exit bot.

Commit:
Add Gradle support

MerlinLim committed:
Add Greet, Echo and Exit

Chat bot starts by greeting the user and echos user's inputs.
When user types, "bye" the bot exits and stops.

MerlinLim committed:
Add text, List text

Chat Bot adds whatever text input to an ArrayList.
When User types "list", Chat Bot will display all the text the
user has entered in chronological order.

MerlinLim committed:
Create Task class and includes done feature

Task class has two attributes, String name and boolean done. markDone
sets done = True and getName gets name. An overriden string method
returns the object's name and a symbol to indicate if the Task has been
done or not done.

The main method now includes the command "done" to allow the user to
mark tasks that are done. E.g "done 1" will mark the first task done.

MerlinLim committed:
Adds todo, deadline and event class

Todo deadline and event class are subclasses of Task.
Todo class is for tasks with no date/time attached
Deadline class is for tasks with a data/time to indicate when the
dealine is.
Event class is for tasks with date/time to indicate when it starts and
when it ends.

Data/time are strings.

The main includes todo, deadline and event features. If user specifies
which type of task to do, the main will handle each type accordingly.

MerlinLim committed:
Adds input/output Automated testing files

Adds automated testing files to the files.

MerlinLim committed:
Adds two exception cases

If todo description is empty, user is prompted to include

If task type is not specified, the user is prompted that the bot does
not know what the user is trying to say.

MerlinLim committed:
Add Delete feature

Adds delete feature. If user specifies delete command, it will remove
the specified task (in terms of index ) from the task list.

MerlinLim committed:
Add class DukeException

Create DukeException and main throw DukeException when invalid command
or todo is missing a description.

MerlinLim committed:
Add save feature and data file

Chat Bot saves data file by parsing each task the follwing format
TaskType/isDone/Desc/Desc_1. Only for todo task, Desc_1 is ommited.

Using bufferedwriter to write the textfile Duke.txt because stream
flushes and closes regularly after each command.

MerlinLim committed:
Add date format to Deadline class

Import LocalDate and DateTimeFormatter to convert string desc
to a LocalDate type.

Convert input "uuuu-MM-dd" to "LL dd uuuu" for deadline tasks.

MerlinLim committed:
Merge branch 'branch-Level-7'

Add save feature

MerlinLim committed:
Merge branch 'branch-Level-8'

Add Date Formatting for Deadline Class.

Resolve merge conflict by adding all import statements.

MerlinLim committed:
Fix Date Bug

Correctly parses the deadline time into data/duke.txt

MerlinLim committed:
Add Ui, Storage, Parser and TaskList classes

The List of classess is as follows:
1. Ui: Deals with interactions with user and manages System output

2. Storage: Deals with loading tasks from duke.txt and writing new
data to the file.

3. Parser: Deals with intepreting user inputs and returning the
parsing it as their respective tasks or modes. enum Modes is used to
differentiate between each command to tell the program what action
is required.

4. TaskList: Encapsulates the list of tasks as an Array List and
contains methods such as add/remove/markDone to modify the queue.

The Task.java file contains the task class and its subclasses.
DukeException file contains the DukeException class.

The main method in Duke.java only contains the logic of the program.

Whleee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,280 @@
+        ArrayList<Task> list = new ArrayList<>();
+
+        //Destination
+        String dest = "data/duke.txt";
```
Perhaps can just use destination as the variable name since it is not too long?

murtubak replied:
I agree, destination is not long enough to need to shorten to dest.

Whleee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,280 @@
+                    System.out.println("Noted. I've removed this task:\n" + removed.toString() +
+                            "\nNow you have " + list.size() + " tasks in the list.");
+                } else {
+                    // Adds item to list
```
Maybe try to keep the formatting coherent, in this case a new line after the else statement.

Whleee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,280 @@
+
+
+
+class Task {
```
Perhaps we can move these classes out of the Duke class for better organisation.

Whleee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,280 @@
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+    public static void main(String[] args){
```
Maybe can look into abstracting out the main logic of Duke into another class to add a layer of abstraction.

Whleee commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,280 @@
+                            name = input.split(" /by ")[0].split("deadline ")[1];
+                            String by = input.split(" /by ")[1];
+                            task = new Deadline(name, by);
+
```
Spacing seems to be inconsistent.

Whleee reviewed (commented):
Completed Review 2.
Looks good overall.
Minor style changes suggested.

MerlinLim committed:
Add packages, move classes to Duke folder

Create Duke folder within java folder and move class files to Duke
folder. Import classess as packages in Duke.java.

Sepearte Task subclasses into their respective files
(e.g todo-> todo.java).

Minor grammar fix in printing total current task(s) as singular where
applicable.

MerlinLim committed:
Add JUnit Tests, replace line seperator with \n

Create JUnit tests to test the behavior of the code.
Add two JUnit Testcases under test folder to check the
system output format.

Replace System.lineSeperator with next line character "\n" so that it
is easier to test for system outputs on JUnit.

murtubak commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,81 @@
+import Duke.DukeException;
```
This class should be put into a package.

murtubak commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,81 @@
+            }
+        } catch(IOException e){
+            System.out.println(e.getMessage());
+        }finally{
```
Should leave a spacing between curly braces and finally.

MerlinLim committed:
Add a Parser test case

Test parser output using JUnit test.

murtubak commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,81 @@
+import Duke.Storage;
+import Duke.TaskList;
+import java.io.IOException;
+
```
Missing Javadocs.

murtubak commented on src/main/java/Duke/Event.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + at + ")";
```
A little inconsistent with the this.at and at. Not that there is anything wrong.

murtubak commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+
+    public Mode mode(String input) throws DukeException {
+        String command = input.split(" ")[0];
+        switch (command) {
+            case "todo":
+                return Mode.TODO;
+            case "deadline":
+                return Mode.DEADLINE;
+            case "event":
+                return Mode.EVENT;
+            case "bye":
+                return Mode.BYE;
+            case "done":
+                return Mode.DONE;
+            case "delete":
+                return Mode.DELETE;
+            case "list":
+                return Mode.LIST;
+            default:
+                throw new DukeException("\u2639 OOPS!!! I'm sorry, but I don't know what that means :-(");
```
Case statements should not have extra indentation. (they should be placed on the same indentation as the switch statement)

murtubak commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+
+    }
+
+    public void writeData(String text, Boolean appendMode) throws IOException{
```
Boolean should be made to sound like a boolean. Perhaps doesWriteData?

murtubak commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,54 @@
+
+    public Ui(){
+        this.sc = new Scanner(System.in);
+    }
+
+    public void showWelcome(){
+        System.out.println("Hello! I'm Duke\n" +
+                "What can I do for you?");
+    }
+
+    public String readLine(){
+        return sc.nextLine();
+    }
+
+    public void close(){
+        this.sc.close();
+    }
+
+    public void goodbye(){
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
+    public void showAddTask(Task task){
+        System.out.println("Go it. I've added this task:\n" + task.toString());
+    }
+
+    public void showDoneTask(Task task){
+        System.out.println("Nice! I've marked this task as done:\n" + task);
+    }
+
+    public void showRemovedTask(Task task){
+        System.out.println("Noted. I've removed this task:\n" + task.toString());
+    }
+
+    public void showTotalTasks(int size){
+        String plural = size != 1 ? "tasks" : "task";
+        System.out.println("Now you have " + size + " " + plural + " in the list.");
+    }
+
+    public void printMessage(String output){
```
Should leave a space before the opening curly braces for the method.

murtubak reviewed (commented):
In general, not that many coding standard violations. I like the well-blocked code chunks that made the code very readable.

MerlinLim committed:
Add relative path references and create data file

Path references to the data file are relative to the source file.

If the destination folder and duke.txt not initialised yet, Storage
will create the folder and text file when it is instantiated.

Remove data folder because Storage class creates it for the user.

MerlinLim committed:
Add JavaDoc comments to most classes

Brief description for just the public methods and classes.

Also fix import statements and some bugs.

MerlinLim committed:
Implemnt coding standard

Change indentation, whitespaces and variable names to comply with
coding standards.

MerlinLim committed:
Add find task feature

Find will search for task names containing the specificed string and
return a list of matching tasks.

User input command "find <string>" will find the
tasks matching <string>

MerlinLim committed:
Merge branch 'branch-A-JavaDoc'

MerlinLim committed:
Merge branch 'branch-A-CodingStandard'

Resolve conflict with merged commit.

MerlinLim committed:
Merge branch 'branch-Level-9'

Resolve conflict with previous commit.
Add 'find' feature to master.

MerlinLim committed:
Merge branch 'add-gradle-support'

Add gradle support.

MerlinLim committed:
Add gradle and TaskList tests

Add gradle feature to automate tests.

Add TaskListTest to test TaskList methods.

MerlinLim committed:
Add GUI feature using JavaFX library

Add Launcher class to launch proj using JavaFX.

Add DialogBox class to handle dialog boxes formattings

Resource folder contains images to be used for the GUI

Two changes to Duke class: Remove main method in Duke,
use launcher main instead to run bot.

Start method ininitialises files, storage and tasklist using method
init before proceeding

Add JavaFX dependencies into build.gradle

MerlinLim committed:
Merge branch 'branch-GUI'

Add GUI feature into bot using JavaFX.

MerlinLim committed:
Add Assertion for task inputs

Assertion ensures task inputs have the correct number of arguments
E.g 3 arg for deadline, 2 for todo.

Remove excess png images.

MerlinLim committed:
Move code into new packages, improve code quality

Add new packages to categorise classes of the same cateogry to navigate
code easily. E.g Task Package contains TaskList, Task, Deadline and etc.

Add MainWindow class to represent the application window and abstract
the formatting and dialog Box creation.

Add Commandhandler to abstract away how the bot handles each command.
E.g Task commands need to be added to tasklist, stored while List
command only needs to return string output. Use public static method
execute to handle the user command.

Create a new file for enum class Mode, under command package.

Duke encapsulates the bot's current TaskList, Storage and exit status.

Use Main.main() to run bot. build.gradle and ReadME.md changes
accordingly

MerlinLim committed:
Merge pull request #1 from MerlinLim/branch-A-Assertions

Add Assertion for task inputs

MerlinLim committed:
Merge branch 'branch-A-CodeQuality'

Organise code into new packages and improve code.

MerlinLim committed:
Add Tagging feature

Able to tag Tasks when using the command tag.
e.g tag 1 favourite will tag the task favourite and be shown in the
list of tasks with a hashtag.

MerlinLim committed:
Merge branch 'branch-C-Tagging'

Add Tagging feature.

MerlinLim committed:
Improve code quality and add JavaDocs

Replace string concatencaction with StringBuilder and String.format.

Fix code indentation and whitespaces.

Add exceptions for invalid deadline input.

MerlinLim committed:
Add Magic Package

Magic package contains all the magic strings the bot needs.
MyString contains Error, symbols, string delimiters and UI responses.

Format contains the strings used for formatting by tasks and UI to
display the tasks to the user.

Fix some bugs.

MerlinLim committed:
Modify README user guide

Include new instructions to setup and sample screenshot

MerlinLim committed:
Change README guide

Add User Guide instructions in docs/README.md and sample photo.

MerlinLim committed:
Fix README and Ui Screenshot

MerlinLim committed:
Fix formatting of sample screen shot

MerlinLim committed:
Format correction to README.md
